src/platform/library/lib/memory.c

```diff
@@ -0,0 +1,234 @@
+static SHARED_DATA struct block_hdr mod_block512[HEAP_RT_COUNT512];
+static SHARED_DATA struct block_hdr mod_block1024[HEAP_RT_COUNT1024];
+static SHARED_DATA struct block_hdr mod_block2048[HEAP_RT_COUNT2048];
+static SHARED_DATA struct block_hdr mod_block4096[HEAP_RT_COUNT4096];
+
+/* Heap memory map for modules */
+static SHARED_DATA struct block_map rt_heap_map[] = {
+	BLOCK_DEF(64, HEAP_RT_COUNT64, uncached_block_hdr(mod_block64)),
+	BLOCK_DEF(128, HEAP_RT_COUNT128, uncached_block_hdr(mod_block128)),
+	BLOCK_DEF(256, HEAP_RT_COUNT256, uncached_block_hdr(mod_block256)),
+	BLOCK_DEF(512, HEAP_RT_COUNT512, uncached_block_hdr(mod_block512)),
+	BLOCK_DEF(1024, HEAP_RT_COUNT1024, uncached_block_hdr(mod_block1024)),
+	BLOCK_DEF(2048, HEAP_RT_COUNT2048, uncached_block_hdr(mod_block2048)),
+	BLOCK_DEF(4096, HEAP_RT_COUNT4096, uncached_block_hdr(mod_block4096)),
+};
+
+#if CONFIG_CORE_COUNT > 1
+/* Heap blocks for runtime shared */
+static SHARED_DATA struct block_hdr rt_shared_block64[HEAP_RUNTIME_SHARED_COUNT64];
+static SHARED_DATA struct block_hdr rt_shared_block128[HEAP_RUNTIME_SHARED_COUNT128];
+static SHARED_DATA struct block_hdr rt_shared_block256[HEAP_RUNTIME_SHARED_COUNT256];
+static SHARED_DATA struct block_hdr rt_shared_block512[HEAP_RUNTIME_SHARED_COUNT512];
+static SHARED_DATA struct block_hdr rt_shared_block1024[HEAP_RUNTIME_SHARED_COUNT1024];
+
+/* Heap memory map for runtime shared */
+static SHARED_DATA struct block_map rt_shared_heap_map[] = {
+	BLOCK_DEF(64, HEAP_RUNTIME_SHARED_COUNT64, uncached_block_hdr(rt_shared_block64)),
+	BLOCK_DEF(128, HEAP_RUNTIME_SHARED_COUNT128, uncached_block_hdr(rt_shared_block128)),
+	BLOCK_DEF(256, HEAP_RUNTIME_SHARED_COUNT256, uncached_block_hdr(rt_shared_block256)),
+	BLOCK_DEF(512, HEAP_RUNTIME_SHARED_COUNT512, uncached_block_hdr(rt_shared_block512)),
+	BLOCK_DEF(1024, HEAP_RUNTIME_SHARED_COUNT1024, uncached_block_hdr(rt_shared_block1024)),
+};
+#endif
+
+/* Heap blocks for buffers */
+static SHARED_DATA struct block_hdr buf_block[HEAP_BUFFER_COUNT];
+static SHARED_DATA struct block_hdr lp_buf_block[HEAP_LP_BUFFER_COUNT];
+
+/* Heap memory map for buffers */
+static SHARED_DATA struct block_map buf_heap_map[] = {
+	BLOCK_DEF(HEAP_BUFFER_BLOCK_SIZE, HEAP_BUFFER_COUNT,
+		  uncached_block_hdr(buf_block)),
+};
+
+static SHARED_DATA struct block_map lp_buf_heap_map[] = {
+	BLOCK_DEF(HEAP_LP_BUFFER_BLOCK_SIZE, HEAP_LP_BUFFER_COUNT,
+		  uncached_block_hdr(lp_buf_block)),
+};
+
+static SHARED_DATA struct mm memmap;
+
+void platform_init_memmap(struct sof *sof)
+{
+	int i;
+
+	/* access memory map through uncached region */
+	sof->memory_map = cache_to_uncache(&memmap);
+
+	/* .system primary core initialization */
+	sof->memory_map->system[0].heap = (unsigned long)malloc(HEAP_SYSTEM_M_SIZE);
+	sof->memory_map->system[0].size = HEAP_SYSTEM_M_SIZE;
+	sof->memory_map->system[0].info.free = HEAP_SYSTEM_M_SIZE;
+	sof->memory_map->system[0].caps = SOF_MEM_CAPS_RAM | SOF_MEM_CAPS_EXT |
+		SOF_MEM_CAPS_CACHE;
```

Comment:
Has `system[0]` other fields already initialized that need to be preserved here? If not then C99 designated initializers would be much more readable:
```C
sof->memory_map->system[0] = {
    .heap = (unsigned long)malloc(HEAP_SYSTEM_M_SIZE);
    ...

```

Same comment several times below.

Reply:
The intention is to leave it the same as cavs map atm to track any changes. This will all be redundant when Zephyr allocation is integrated.